File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    const graph = {};
    
    // 1. 그래프 구성
    for (const [from, to] of tickets) {
        if (!graph[from]) graph[from] = [];
        graph[from].push(to);
    }
    
    // 2. 사전순 정렬
    for (const key in graph) {
        graph[key].sort();
    }
    
    const route = [];
    
    // 3. DFS (Hierholzer)
    function dfs(airport) {
        while (graph[airport] && graph[airport].length > 0) {
            const next = graph[airport].shift(); // 사전순
            dfs(next);
        }
        route.push(airport);
    }
    
    dfs("ICN");
    
    // 4. 역순이 정답
    return route.reverse();
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    
    visited = [False] * len(tickets)
    
    def dfs(cur):
        if len(cur) == len(tickets) + 1:
 # 주의: append(cur)을 하면 같은 리스트 객체를 참조로 넣어두는 거라 DFS가 끝나고 pop() 되면 answer 안에 들어간 값도 같이 바뀜
            answer.append(cur[:])          
            return
        
        for i in range(len(tickets)):
            start, end = tickets[i]
            if not visited[i] and cur[-1] == start:
                
                visited[i] = True
                cur.append(end)
                dfs(cur)
                cur.pop()
                visited[i] = False
        
    dfs(['ICN'])
    
    answer.sort()
    
    return answer[0]